GatewayIndexStateIT#testJustMasterNode fails

e.g. https://scans.gradle.com/s/6lmw3lvxj7ibc/console-log?task=:server:integTest failed like this:

```
org.elasticsearch.gateway.GatewayIndexStateIT > testJustMasterNode FAILED
    java.lang.AssertionError: 
    Expected: <true>
         but: was <false>
        at __randomizedtesting.SeedInfo.seed([3611F64084CD65A4:93E62E9ADE828DC8]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.gateway.GatewayIndexStateIT.testJustMasterNode(GatewayIndexStateIT.java:236)
REPRODUCE WITH: ./gradlew :server:integTest --tests "org.elasticsearch.gateway.GatewayIndexStateIT.testJustMasterNode" -Dtests.seed=3611F64084CD65A4 -Dtests.security.manager=true -Dtests.locale=sr-Latn-ME -Dtests.timezone=Africa/Monrovia -Dcompiler.java=12 -Druntime.java=8
```

Relevant logs:

```
  1> [2019-07-16T06:15:43,243][INFO ][o.e.g.GatewayIndexStateIT] [testJustMasterNode] --> starting 1 master node non data again
  1> [2019-07-16T06:15:43,247][INFO ][o.e.e.NodeEnvironment    ] [testJustMasterNode] using [1] data paths, mounts [[/ (rootfs)]], net usable_space [302.5gb], net total_space [349.9gb], types [rootfs]
  1> [2019-07-16T06:15:43,247][INFO ][o.e.e.NodeEnvironment    ] [testJustMasterNode] heap size [491mb], compressed ordinary object pointers [true]
  1> [2019-07-16T06:15:43,250][INFO ][o.e.n.Node               ] [testJustMasterNode] node name [node_t0], node ID [a3ZIV7VzRh2mGGoFVkcufw], cluster name [TEST-TEST_WORKER_VM=[79]-CLUSTER_SEED=[-3993295015402633431]-HASH=[16CAED26BA4]-cluster]
  1> [2019-07-16T06:15:43,250][INFO ][o.e.n.Node               ] [testJustMasterNode] version[7.4.0], pid[47282], build[unknown/unknown/Unknown/Unknown], OS[Linux/3.10.0-957.21.3.el7.x86_64/amd64], JVM[Oracle Corporation/Java HotSpot(TM) 64-Bit Server VM/1.8.0_212/25.212-b10]
  1> [2019-07-16T06:15:43,250][INFO ][o.e.n.Node               ] [testJustMasterNode] JVM home [/var/lib/jenkins/.java/oracle-8u212-linux/jre]
  1> [2019-07-16T06:15:43,250][INFO ][o.e.n.Node               ] [testJustMasterNode] JVM arguments [-Dfile.encoding=UTF8, -Dcompiler.java=12, -Des.scripting.update.ctx_in_params=false, -Dgradle.dist.lib=/var/lib/jenkins/.gradle/wrapper/dists/gradle-5.5-all/byeise1led9a3w3p149d02kfj/gradle-5.5/lib, -Dgradle.user.home=/var/lib/jenkins/.gradle, -Dgradle.worker.jar=/var/lib/jenkins/.gradle/caches/5.5/workerMain/gradle-worker.jar, -Djava.awt.headless=true, -Djava.security.manager=worker.org.gradle.process.internal.worker.child.BootstrapSecurityManager, -Djna.nosys=true, -Dorg.gradle.native=false, -Druntime.java=8, -Dtests.artifact=server, -Dtests.gradle=true, -Dtests.logger.level=WARN, -Dtests.security.manager=true, -Dtests.seed=3611F64084CD65A4, -Dtests.task=:server:integTest, -XX:+HeapDumpOnOutOfMemoryError, -XX:HeapDumpPath=/var/lib/jenkins/workspace/elastic+elasticsearch+7.x+multijob+fast+part1/server/build/heapdump, -esa, -Xms512m, -Xmx512m, -Dfile.encoding=UTF-8, -Djava.io.tmpdir=./temp, -Duser.country=US, -Duser.language=en, -Duser.variant, -ea]
  1> [2019-07-16T06:15:43,251][INFO ][o.e.p.PluginsService     ] [testJustMasterNode] no modules loaded
  1> [2019-07-16T06:15:43,251][INFO ][o.e.p.PluginsService     ] [testJustMasterNode] loaded plugin [org.elasticsearch.test.ESIntegTestCase$AssertActionNamePlugin]
  1> [2019-07-16T06:15:43,251][INFO ][o.e.p.PluginsService     ] [testJustMasterNode] loaded plugin [org.elasticsearch.test.ESIntegTestCase$TestSeedPlugin]
  1> [2019-07-16T06:15:43,251][INFO ][o.e.p.PluginsService     ] [testJustMasterNode] loaded plugin [org.elasticsearch.test.MockHttpTransport$TestPlugin]
  1> [2019-07-16T06:15:43,251][INFO ][o.e.p.PluginsService     ] [testJustMasterNode] loaded plugin [org.elasticsearch.transport.nio.MockNioTransportPlugin]
  1> [2019-07-16T06:15:43,264][INFO ][o.e.d.DiscoveryModule    ] [testJustMasterNode] using discovery type [zen] and seed hosts providers [settings, file]
  1> [2019-07-16T06:15:43,276][INFO ][o.e.n.Node               ] [testJustMasterNode] initialized
  1> [2019-07-16T06:15:43,277][INFO ][o.e.n.Node               ] [[test_TEST-TEST_WORKER_VM=[79]-CLUSTER_SEED=[-3993295015402633431]-HASH=[16CAED26BA4]-cluster[T#2]]] starting ...
  1> [2019-07-16T06:15:43,281][INFO ][o.e.t.TransportService   ] [[test_TEST-TEST_WORKER_VM=[79]-CLUSTER_SEED=[-3993295015402633431]-HASH=[16CAED26BA4]-cluster[T#2]]] publish_address {127.0.0.1:44830}, bound_addresses {[::1]:40828}, {127.0.0.1:44830}
  1> [2019-07-16T06:15:43,282][INFO ][o.e.c.c.Coordinator      ] [[test_TEST-TEST_WORKER_VM=[79]-CLUSTER_SEED=[-3993295015402633431]-HASH=[16CAED26BA4]-cluster[T#2]]] cluster UUID [pjq-ik91Tra78fExUyo-1A]
  1> [2019-07-16T06:15:43,354][INFO ][o.e.c.s.MasterService    ] [node_t0] elected-as-master ([1] nodes joined)[{node_t0}{a3ZIV7VzRh2mGGoFVkcufw}{-mwzDtCQQ6ufccSkKd8Igw}{127.0.0.1}{127.0.0.1:44830}{im} elect leader, _BECOME_MASTER_TASK_, _FINISH_ELECTION_], term: 2, version: 5, reason: master node changed {previous [], current [{node_t0}{a3ZIV7VzRh2mGGoFVkcufw}{-mwzDtCQQ6ufccSkKd8Igw}{127.0.0.1}{127.0.0.1:44830}{im}]}
  1> [2019-07-16T06:15:43,359][INFO ][o.e.c.s.ClusterApplierService] [node_t0] master node changed {previous [], current [{node_t0}{a3ZIV7VzRh2mGGoFVkcufw}{-mwzDtCQQ6ufccSkKd8Igw}{127.0.0.1}{127.0.0.1:44830}{im}]}, term: 2, version: 5, reason: Publication{term=2, version=5}
  1> [2019-07-16T06:15:43,360][INFO ][o.e.n.Node               ] [[test_TEST-TEST_WORKER_VM=[79]-CLUSTER_SEED=[-3993295015402633431]-HASH=[16CAED26BA4]-cluster[T#2]]] started
  1> [2019-07-16T06:15:43,360][INFO ][o.e.g.GatewayIndexStateIT] [testJustMasterNode] --> waiting for test index to be created
  1> [2019-07-16T06:15:43,361][INFO ][o.e.g.GatewayIndexStateIT] [testJustMasterNode] --> verify we have an index
  1> [2019-07-16T06:15:43,361][INFO ][o.e.g.GatewayIndexStateIT] [testJustMasterNode] [GatewayIndexStateIT#testJustMasterNode]: cleaning up after test
  1> [2019-07-16T06:15:43,364][INFO ][o.e.g.GatewayService     ] [node_t0] recovered [1] indices into cluster_state
  1> [2019-07-16T06:15:43,364][INFO ][o.e.c.m.MetaDataDeleteIndexService] [node_t0] [test/Gi1twbiPQHO1rOGvMZPEFQ] deleting index
  1> [2019-07-16T06:15:43,369][INFO ][o.e.n.Node               ] [testJustMasterNode] stopping ...
  1> [2019-07-16T06:15:43,371][INFO ][o.e.n.Node               ] [testJustMasterNode] stopped
```

Note that `waiting for test index to be created` completes before `recovered [1] indices into cluster_state` is logged. I suspect this is due to a change in how cluster health works in #44348.

`REPRODUCE LINE` does not reproduce, nor do initial attempts with large numbers of iterations.

--- a/server/src/main/java/org/elasticsearch/action/admin/cluster/health/TransportClusterHealthAction.java
+++ b/server/src/main/java/org/elasticsearch/action/admin/cluster/health/TransportClusterHealthAction.java
@@ -138,7 +138,13 @@ public class TransportClusterHealthAction extends StreamableTransportMasterNodeR
                         final long timeoutInMillis = Math.max(0, endTimeRelativeMillis - threadPool.relativeTimeInMillis());
                         final TimeValue newTimeout = TimeValue.timeValueMillis(timeoutInMillis);
                         request.timeout(newTimeout);
-                        executeHealth(request, newState, listener, waitCount,
+
+                        // we must use the state from the applier service, because if the state-not-recovered block is in place then the
+                        // applier service has a different view of the cluster state from the one supplied here
+                        final ClusterState appliedState = clusterService.state();
+                        assert newState.stateUUID().equals(appliedState.stateUUID())
+                            : newState.stateUUID() + " vs " + appliedState.stateUUID();
+                        executeHealth(request, appliedState, listener, waitCount,
                             observedState -> waitForEventsAndExecuteHealth(request, listener, waitCount, endTimeRelativeMillis));
                     }
 

--- a/server/src/test/java/org/elasticsearch/gateway/GatewayIndexStateIT.java
+++ b/server/src/test/java/org/elasticsearch/gateway/GatewayIndexStateIT.java
@@ -211,7 +211,6 @@ public class GatewayIndexStateIT extends ESIntegTestCase {
         client().prepareIndex("test", "type1", "2").setSource("field1", "value1").execute().actionGet();
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/44416")
     public void testJustMasterNode() throws Exception {
         logger.info("--> cleaning nodes");
 
@@ -221,11 +220,13 @@ public class GatewayIndexStateIT extends ESIntegTestCase {
         logger.info("--> create an index");
         client().admin().indices().prepareCreate("test").setWaitForActiveShards(ActiveShardCount.NONE).execute().actionGet();
 
-        logger.info("--> closing master node");
-        internalCluster().closeNonSharedNodes(false);
-
-        logger.info("--> starting 1 master node non data again");
-        internalCluster().startNode(Settings.builder().put(Node.NODE_DATA_SETTING.getKey(), false).build());
+        logger.info("--> restarting master node");
+        internalCluster().fullRestart(new RestartCallback(){
+            @Override
+            public Settings onNodeStopped(String nodeName) {
+                return Settings.builder().put(Node.NODE_DATA_SETTING.getKey(), false).build();
+            }
+        });
 
         logger.info("--> waiting for test index to be created");
         ClusterHealthResponse health = client().admin().cluster().prepareHealth().setWaitForEvents(Priority.LANGUID).setIndices("test")
@@ -237,7 +238,7 @@ public class GatewayIndexStateIT extends ESIntegTestCase {
         assertThat(clusterStateResponse.getState().metaData().hasIndex("test"), equalTo(true));
     }
 
-    public void testJustMasterNodeAndJustDataNode() throws Exception {
+    public void testJustMasterNodeAndJustDataNode() {
         logger.info("--> cleaning nodes");
 
         logger.info("--> starting 1 master node non data");

